doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,487 @@
+                                                 void *arg),
+                                      void *arg);
+int BIO_write_queue(BIO *b, const void *buf, size_t buf_len);
+int BIO_flush(BIO *b);
+```
+
+The status argument to the write done callback will be 1 on success, some
+negative value on failure, and some special negative value if the BIO_dgram is
+being freed before the write could be completed.
+
+For send/receive addresses, we import the `BIO_(set|get)_dgram_(origin|dest)`
+APIs proposed in the sendmsg/recvmsg PR (#5257). `BIO_get_dgram_(origin|dest)`
+should be called immediately after `BIO_read_dequeue` and
+`BIO_set_dgram_(origin|dest)` should be called immediately before
+`BIO_write_queue`.
+
+This approach allows `BIO_dgram` to support myriad options via composition of
+successive function calls in a “builder” style rather than via a single function
+call with an excessive number of arguments or pointers to unwieldy ever-growing
+argument structures, requiring constant revision of the central read/write
+functions of the BIO API.
+
+Note that since `BIO_set_dgram_(origin|dest)` sets data on outgoing packets and
+`BIO_get_dgram_(origin|dest)` gets data on incoming packets, it doesn't follow
+that these are accessing the same data (they are not setters and getters of a
+variables called "dgram origin" and "dgram destination", even though they look
+like setters and getters of the same variables from the name.) We probably want
+to separate these as there is no need for a getter for outgoing packet
+destination, for example, and by separating these we allow the possibility of
+multithreaded use (one thread reads, one thread writes) in the future. Possibly
+we should choose less confusing names for these functions. Maybe
+`BIO_set_outgoing_dgram_(origin|dest)` and
+`BIO_get_incoming_dgram_(origin|dest)`.
+
+Pros of this approach:
+
+  - Application can generate one datagram at a time and still get the advantages
+    of sendmmsg/recvmmsg (fewer syscalls, etc.)
+
+    We probably want this for our own QUIC implementation built on top of this
+    anyway. Otherwise we will need another piece to do basically the same thing
+    and agglomerate multiple datagrams into a single BIO call. Unless we only
+    want use `sendmmsg` constructively in trivial cases (e.g. where we send two
+    datagrams from the same function immediately after one another... doesn't
+    seem like a common use case.)
+
+  - Flexible support for single-copy (zero-copy).
+
+Cons of this approach:
+
+  - Very different way of doing reads/writes might be strange to existing
+    applications. *But* the primary consumer of this new API will be our own
+    QUIC implementation so probably not a big deal. We can always support
+    `BIO_read`/`BIO_write` as a less efficient fallback for existing third party
+    users of BIO_dgram.
+
+### Compatibility interop
+
+Suppose the following sequence happens:
+
+1. BIO_read (legacy call path)
```

Comment:
I'd be inclined to ignore such intermixing.  No existing application can do it and we should be able to police it adequately.